client/components/FacetsStateProvider/actions/updateFacet.ts

```diff
@@ -0,0 +1,128 @@
+import { GetState, SetState } from 'zustand';
+
+import {
+	FacetDefinition,
+	FacetInstanceStack,
+	FacetInstanceType,
+	FacetSourceScope,
+	FacetName,
+	Facets,
+	cascade,
+	parsePartialFacetInstance,
+	createFacetInstance,
+	FacetCascadeResult,
+} from 'facets';
+
+import { FacetsState, FacetState } from '../types';
+
+function applyPatchToStack<Def extends FacetDefinition>(
+	def: Def,
+	stack: FacetInstanceStack<Def>,
+	patch: Partial<FacetInstanceType<Def>>,
+	scope: FacetSourceScope,
+): FacetInstanceStack<Def> {
+	const patchableIndex = stack.findIndex(
+		(e) => e.scope.kind === scope.kind && e.scope.id === scope.id,
+	);
+	if (patchableIndex >= 0) {
+		const entry = stack[patchableIndex];
+		const { value } = entry;
+		const patchedEntry = {
+			...entry,
+			value: { ...value, ...patch },
+		};
+		const nextStack = stack.concat();
+		nextStack.splice(patchableIndex, 1, patchedEntry);
+		return nextStack;
+	}
+	return [...stack, { scope, facetBindingId: null, value: createFacetInstance(def, patch) }];
```

Comment:
FFTI, thoughts on using immer here in some of these updater functions to simplify? I'm still on the fence about immer myself.

Reply:
I'd be in favor of a future refactor using Immer!